### PR TITLE
Support SO_REUSEPORT option

### PR DESCRIPTION
**Pull request by zimathias**
so that multiple osrm-routed processes can bind to the same port.

### CODE REVIEW

File: include/server/server.hpp [hunk 1]
```diff
@@ -49,7 +49,12 @@ class Server
         boost::asio::ip::tcp::endpoint endpoint = *resolver.resolve(query);
 
         acceptor.open(endpoint.protocol());
+#ifdef SO_REUSEPORT
+        int one = 1;
```
Comment (daniel-j-h) [inline include/server/server.hpp]: Can be const?

File: include/server/server.hpp [hunk 2]
```diff
@@ -49,7 +49,12 @@ class Server
...
+        setsockopt(acceptor.native_handle(), SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &one, sizeof(one));
```
Comment (daniel-j-h) [inline include/server/server.hpp]: This relies on the particular Boost.ASIO header we're including to transitively include sys/types.h, sys/socket.h.

**Comment by zimathias** [comment]
@daniel-j-h let me know if this PR addresses your comments now.

**Comment by daniel-j-h** [comment]
The unconditional includes currently break Appveyor, I would just ifdef them.

File: include/server/server.hpp [hunk 3]
```diff
@@ -13,6 +13,9 @@
 
 #include <zlib.h>
 
+#include <sys/types.h>
```
Comment (TheMarex) [inline include/server/server.hpp]: needs `#ifndef _WIN32` around this since `sys/` is platform specific.

**Comment by TheMarex** [comment]
I'm curious why do you use multiple processes over multiple threads?

Can you rebase this on `master`? There seem to be a few stray merge conflicts in your branch.

**Comment by zimathias** [comment]
@TheMarex this allows to hot reload the osrm-routed binary without dropping requests. Given SO_REUSEPORT allows multiple processes to bind to the same port the binary hot reload process becomes:
1. start new process with new binary
2. wait for the new process to startup and serve requests
3. stop the old process with the old binary

**Comment by TheMarex** [comment]
Closing in favor of https://github.com/Project-OSRM/osrm-backend/pull/2417

**Pull request closed without merging**